### PR TITLE
Contractible property

### PR DESCRIPTION
**Pull request by GeoffreySangston**
I added P000199 (Contractible), T000564 (Contractible => Path Connected), and T000565 (Indiscrete => Contractible).
I also added P000199 (Contractible) as a property to S000025 (Euclidean Real Numbers).

> @GeoffreySangston This is a perfect size for an initial PR.
> 
> Can you change the PR title to something more specific (e.g., "Contractible property")? That will make it easier to figure what each PR is about when scanning the list (in particular the list of closed PRs later on).

I thought I had to delete the branch (because I thought the pull request title might be the same as the branch name) and submit a new pull request. The old pull request is discussed here:
https://github.com/pi-base/data/pull/830

I am not sure now. I apologize if that was unnecessary.

### CODE REVIEW

**Comment by danflapjax** [comment]
For future reference, you can change the name of a PR after you created it by clicking the Edit button at the top of this page. No worries though, it's not a big deal.

**Comment by prabau** [comment]
Thanks for your first contribution.  We appreciate it.

May we ask what your level of experience is with git/github?  Apologies if you are already an expert.  If not, feel free to ask for help on how to do something any time.  Also, which editor are you using?

Now some general comments.
- The last line of each file should be terminated by a newline (or CR/LF on Windows), i.e. need to press the Return key on the keyboard.
- In the `refs:` section, the `name:` part of each reference should be the name of the book or title of the paper/mathse ref, etc.  That was done properly for P199, but not for T564, T565.

- T564, T565 are also being used in a pending PR by @danflapjax, so one of them will have to change when the time comes to merge.  You can leave yours as is for now, and we can change later if needed.  Just FYI.

**Comment by prabau** [comment]
And some comments specific to the contents here.

P199: Your definition is complete.  I think it would also be useful to first mention the more conceptual version, namely that the identity map $X\to X$ is homotopic to some constant map on $X$.  And follow up with what you have.
(See Willard 32.6.)
One could also reference https://en.wikipedia.org/wiki/Contractible_space if you think it's useful.

T564: The exercise from Munkres does not give any details.  Note: it's also Problem 32A from Willard.  Instead you an give a direct proof, which is pretty easy.  Or even better, most probably someone has already asked this on mathse.  Is there a good post about this on mathse that we can just quote?

T565: not sure about this.

Questions: Is the empty set indiscrete?  Is the empty set contractible?
The reason I am asking is that these theorems are then used for automated deduction of results and traits for spaces. So all the edge cases need to be handled properly.

**Comment by GeoffreySangston** [comment]
> Thanks for your first contribution. We appreciate it.
> 
> May we ask what your level of experience is with git/github? Apologies if you are already an expert. If not, feel free to ask for help on how to do something any time. Also, which editor are you using?
> 

I'm certainly not very knowledgeable with git generally, though I have used it via terminal commands for my own personal projects over roughly 10 years (and in a professional capacity a long time ago). The issue is I've generally just adopted fairly simple workflows and never explored it very much (it's pretty easy to do a few basic things a huge number of times), so I'm very much open to helpful critique. I could benefit by rereading some of the basics. I tend to do everything with vim, but for these commits I just used the online gitdev editor.

> Now some general comments.
> 
> * The last line of each file should be terminated by a newline (or CR/LF on Windows), i.e. need to press the Return key on the keyboard.

I'll remember this. Thanks!

> * In the `refs:` section, the `name:` part of each reference should be the name of the book or title of the paper/mathse ref, etc.  That was done properly for P199, but not for T564, T565.

> * T564, T565 are also being used in a pending PR by @danflapjax, so one of them will have to change when the time comes to merge.  You can leave yours as is for now, and we can change later if needed.  Just FYI.

Got it!

**Comment by prabau** [comment]
You seem to have enough knowledge of git to do the basic stuff.  I used to use vim as well.  But for working on pi-base, I use vscode on Windows, or github.dev on the web (same thing really).  It has convenient pi-base specific extensions that expand the id's like S45, P123, etc automatically in preview mode.  That's pretty convenient to detect incorrect values.

As for the git workflow, we follow a common model of working in "topic branches" off of `main`, that get reviewed and merged back into `main`.  So we keep making changes on the same topic branch until everybody is satisfied. There is absolutely no need to create separate branches for each additional change while this is being worked on.

For example, changing the title of the PR is completely independent of git branches and commits.  No branch needed for that.  You also ask if you should create a new PR for modifying P199.  Answer: No.  Keep working on it by adding additional commits to the same topic branch.

**Comment by prabau** [comment]
> (Throughout this longish comment I agree that I made some mistakes (see the comment). Should I go ahead and try to fix them with a new pull request?)

It's perfectly fine to make procedural mistakes until the recommended procedure becomes a habit.  No need to apologize!

But absolutely do **not** fix anything with additional pull requests.  Keep adding new commits as part of the same PR (and topic branch).

**Comment by prabau** [comment]
> > T564: The exercise from Munkres does not give any details. Note: it's also Problem 32A from Willard. Instead you an give a direct proof, which is pretty easy. Or even better, most probably someone has already asked this on mathse. Is there a good post about this on mathse that we can just quote?
> 
> I didn't immediately see any really authoritative looking proof. I can look again later. I'll also look through the textbooks.
> 
> Or if those turn up nothing, please feel free to criticize the following.
> 
> Suppose X is contractible and let x 0 , x 1 ∈ X . By assumption there exists some point p ∈ X and a continuous map F : X × [ 0 , 1 ] → X such that F ( x , 0 ) = x and F ( x , 1 ) = p for all x . The restrictions of F to x 0 × [ 0 , 1 ] and F to x 1 × [ 0 , 1 ] determine paths σ 0 , σ 1 : [ 0 , 1 ] → X . Obtain a path from x 0 to x 1 by concatenating σ 0 to the reverse of σ 1 .
> 
> (This is unsatisfying though because I don't know how to make it very precise without make it kind of tedious.)

Yeah, making it too precise would make it tedious and one loses track of the main argument.  One could just say for example that given the continuous map $F:X\times[0,1]\to X$ such that ..., every point $x\in X$ can be joined by a path to $p$ (using the path $t\mapsto F(x,t)$ ). (If some trivial step can be mentally completed by the reader, e.g., the concatenation part, I think it's good enough.)

But easier: quote https://math.stackexchange.com/questions/3959792/contractible-spaces-are-path-connected-proof

Or https://math.stackexchange.com/questions/715720/a-contractible-space-is-path-connected

**Comment by prabau** [comment]
PR #824 from @danflapjax is close to being merged (just pending a few review changes from him).  So to avoid numbering conflicts, can you change your theorem numbers to T583, T584?
(if doing it in github.dev, right click on the file and pick "Rename" in the pop-up menu.  And also change the `uid` in file itself)

**Comment by GeoffreySangston** [comment]
"I added a proof of P199.

I added a link to a mathse answer for a proof of P199."

Sorry again. This should have been T583.

**Comment by GeoffreySangston** [comment]
In the Youtube video tutorial (https://www.youtube.com/watch?v=-us3rzXjGFk) at around 4:05 it's mentioned that there's a way to preview your work. I think Steven's video profile is obscuring the button which does this and I haven't been able to figure out how to preview what I've done.

Is the window with the following information supposed to serve as a preview? How do I get it to preview what I've written?

> compile
This repository is used to collaborate on the underlying data powering the [pi-Base community database of topological spaces](https://topology.pi-base.org/).
>
>To learn more, visit our [wiki](https://github.com/pi-base/data/wiki).
>
>Copyright 2014-2023 Steven Clontz and James Dabbs, [licensed for free public use](https://github.com/pi-base/data/blob/main/LICENSE.md).

**Comment by prabau** [comment]
I will clone your branch to the main repository to enable a preview.

For now you cannot preview the work yourself.  Once you get a few PRs successfully reviewed and merged to get the hang of things, we'll give you the ability to create branches on the main repository itself.

**Comment by prabau** [comment]
Use the `contractible` branch, which you can enter in the Advanced tab on the pi-base web site.

**Comment by prabau** [comment]
One general procedural thing.  You can keep the title of each commit short and to the point.  No need to give the details of every little thing that you did.  That's what looking at the diff is for.  (So here, the commit title ""I added an extra line at bottom of each file." is misleading because there are many other things in the commit.  I could have been just "Changes based on review" for example.  But again, all that does not matter at all.  Nobody will care about the little intermediate changes.  What matters is the end result of each PR.)

Now let me review the contents.

**Comment by prabau** [comment]
Sorry for being picky, but here are a few comments.  Hope I won't have to give that level of comments next time.

P199: You will have seen that properties in pi-base are sometimes given a few equivalent definitions.  Something like:

> blah blah A
> 
> Equivalently, blah blah B.
> 
> Equivalently, blah blah C.

But when doing this, the formulations A, B, C are equivalent but distinct.  In your case, C is not just equivalent to A, it's the same thing, i.e. an exact restatement of A.  So they should be combined.

Also, need to say what identity map you are talking about.  I know it may be obvious, but someone new to the topic may be confused.  We often use the implicit convention that $X$ denotes the space having the property.
I would write something like:

"The identity map $X\to X$ is *null-homotopic*, i.e., homotopic to a constant map.  In other words, there is a point $p\in X$ and a continuous map $F ...$ etc."

If you want to keep the "homotopy equivalent to a point" part, would it be possible to define what that means?  And add some reference to show the equivalence?  (I found several mathse posts about that.)

(Note: Munkres is fine, has a lot of good things, but may not be the most authoritative source for terminology.  It was written for an undergraduate audience and sometimes has idiosyncratic terminology.)  What would you recommend for a precise definition of "homotopy equivalence"?

**Comment by prabau** [comment]
T583: No need for so much stuff.  The mathse post is all we need in this case.  It has a lot of details and multiple answers if anybody cares.  But why did you reference just one answer (mathse: 2923387) instead of the whole post (mathse: 715720)?

T584 is fine now.

Currently pi-base does not know yet whether the empty space is contractible or not.  It's worth adding the trivial implication:

T585: [contractible ==> not empty].
The justification can be a simple "By definition."

**Comment by danflapjax** [comment]
> "The identity map X → X is _null-homotopic_, i.e., homotopic to a constant map. In other words, there is a point p ∈ X and a continuous map F . . . etc."
> 
> If you want to keep the "homotopy equivalent to a point" part, would it be possible to define what that means? And add some reference to show the equivalence? (I found several mathse posts about that.)

These are *not* equivalent though, as they differ in the case of the empty space: every map from the empty space is constant, but the space fails to contain a point.

**Comment by danflapjax** [comment]
I suggest a definition along the lines of "The identity map on $X$ is *null-homotopic*, that is, there exists a continuous homotopy $H:X\times I\rightarrow X$ such that..."

You can include the equivalent variations on the definition by stating "Equivalently, if X is nonempty, ..."

**Comment by GeoffreySangston** [comment]
> 
> These are _not_ equivalent though, as they differ in the case of the empty space: every map from the empty space is constant, but the space fails to contain a point.

Funnily enough it looks like Wikipedia gets this wrong.
> "It follows from these definitions that a space X is contractible if and only if the identity map from X to itself—which is always a homotopy equivalence—is null-homotopic."

https://en.wikipedia.org/wiki/Homotopy#Null-homotopy

**Comment by prabau** [comment]
@danflapjax Actually, every map from the empty set to any set is the empty map. But why is the empty map constant?
Or what should we understand by "constant map"?

**Comment by GeoffreySangston** [comment]
It could mean either:
"There exists a value y such that for all x, f(x) = y"
(This would identity map on empty space is not constant.)

Or

"For all x, y, f(x) = f(y)"

Meaning empty space identity is constant.

**Comment by StevenClontz** [comment]
Is $0^0=0$ or $0^0=1$? 🙃

I'm unsure we should spend much time deliberating something that only affects the empty space; if it's not arbitrary and we make the wrong call, it can be corrected later.

**Comment by danflapjax** [comment]
To me, "for all x,y f(x)=f(y)" is the standard definition of a constant map. *Specifically* excluding the identity function on the empty set feels a little bizarre to me, but it's also amusingly hard to come up with a property of constant functions that depends on that single case. I thought the page from nLab (my go-to source for mathematical pedantry) on [constant morphisms](https://ncatlab.org/nlab/show/constant+morphism) gave me a good insight into the distinction.

Anyways, I don't especially care whether or not the empty space is contractible, but the definition to me reads as self-contradictory as it stands and needs clarification. For a point of comparison, I would like to point out that some sources do not consider the empty space to be path-connected (as it admits no paths), while others, as well as Pi-Base, do (as there exists a path between any two points). We just need to opt for a definition, and I simply wanted to share that when I apply the one from Munkres to the empty space, I conclude that it is contractible.

**Comment by GeoffreySangston** [comment]
> T583: No need for so much stuff. The mathse post is all we need in this case. It has a lot of details and multiple answers if anybody cares. But why did you reference just one answer (mathse: 2923387) instead of the whole post (mathse: 715720)?

Having lots of details and multiple answers seemed like a negative to me, since that leaves room for ambiguity. That one answer (essentially begins) with:
> *Proof:* Suppose that $X$ is contractible. Choose any two points $a, b \in X$. We will construct a path between $a$ and $b$. 

A proof of "contractible => path connected" is what I wanted the viewer to see when they clicked the link.

**Comment by prabau** [comment]
I think we all agree that the empty space is a degenerate case.  Whether it considered contractible or not is not very important and depends on what definition is chosen.

On the other hand, let's take a look at how *contractible space* is defined in the literature. (Munkres is **not** necessarily an authoritative source for terminology.  It was written for an undergraduate audience with often idiosyncratic choices.)

- https://encyclopediaofmath.org/wiki/Contractible_space: $X$ is "homotopy equivalent to a one-point space". (referencing Spanier as a source)
- Hatcher (Algebraic Topology): "A space having the homotopy type of a point."  "This amounts to requiring that the identity map of the space be nullhomotopic, that is, homotopic to a constant map."
- Brown (Topology and Groupoids): "$X$ is is of the homotopy type of a space with only one point."
- May (A Concise Course in Algebraic Topology): "A space $X$ is said to be contractible if it is homotopy equivalent to a point."
- Encyclopedia of General Topology (section k-07): "A space is said to be contractible to a point (or simply a contractible space) if some point is a deformation retract of it."

So every single one of these authoritative sources for algebraic topology require a contractible space to be non-empty.  So it seems that would be the right choice here.  And if the space is non-empty, it's equivalent to some other formulations as discussed involving a specific point $x\in X$, etc.

**Comment by prabau** [comment]
@GeoffreySangston FYI, when you write items in curly braces for automated expansion like `{S000163}`, you can just write `{S163}` without the leading zeros.  Easier to read and works just as well.

**Comment by GeoffreySangston** [comment]
I'll wait for others to comment, but that's convincing to me.

I noticed that Spanier's definition of contractible on page 25 is ambiguous, but in the proof of Theorem 10 (contractible <=> homotopy equivalent to 1) he assumes contractible and gets a contraction to a point.


I did ad hoc google scholar searches for various key words. Maybe this suggests that "contractible => non-empty" is more common?

["contractible space is non-empty"](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C47&q=%22contractible+space+is+non-empty%22&btnG=). This is a biased search, but it does inform us that two papers state this explicitly. I didn't know how to do the biased search which would find papers asserting the opposite.

I searched for ["empty * contractible"](https://scholar.google.com/scholar?start=0&q=%22empty+*+contractible%22&hl=en&as_sdt=0,47) (1,070 results) and found the following two phrases occur a lot. This is just a rough estimate since not every paper is describing topological spaces, but a lot of them are.

["non-empty and contractible"](https://scholar.google.com/scholar?start=110&q=%22non-empty+and+contractible%22&hl=en&as_sdt=0,47) (279 results)
 - I think this phrase suggests the author either accepts 'empty => contractible', or is at least concerned that this is a point of ambiguity. Otherwise the phrase is redundant.


["empty or contractible"](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C47&q=%22empty+or+contractible%22&btnG=) (742 results)
 - I think this suggest that the author either does not accept 'empty => contractible', or is at least concerned that this is a point of ambiguity. Otherwise the phrase is redundant.

Also,
["empty or non-contractible"](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C47&q=%22empty+or+non-contractible%22&btnG=) comes up in 10 results (mostly by the same few authors) in reference to something called the disk complex. I don't know is this is a topological space though and did not look very closely, but I thought it might be worth mentioning.

**Comment by GeoffreySangston** [comment]
I'll wait for others to comment, but that's convincing to me.

I noticed that Spanier's definition of contractible on page 25 is ambiguous, but in the proof of Theorem 10 (contractible <=> homotopy equivalent to 1) he assumes contractible and gets a contraction to a point.


I did ad hoc google scholar searches for various key words. Maybe this suggests that "contractible => non-empty" is more common?

["contractible space is non-empty"](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C47&q=%22contractible+space+is+non-empty%22&btnG=). This is a biased search, but it does inform us that two papers state this explicitly. I didn't know how to do the biased search which would find papers asserting the opposite.

I searched for ["empty * contractible"](https://scholar.google.com/scholar?start=0&q=%22empty+*+contractible%22&hl=en&as_sdt=0,47) (1,070 results) and found the following two phrases occur a lot. This is just a rough estimate since not every paper is describing topological spaces, but a lot of them are.

["non-empty and contractible"](https://scholar.google.com/scholar?start=110&q=%22non-empty+and+contractible%22&hl=en&as_sdt=0,47) (279 results)
 - I think this phrase suggests the author either accepts that empty spaces are contractible, or is at least concerned that this is a point of ambiguity. Otherwise the phrase is redundant.
 - E.g., [*Sum of Hamiltonian manifolds*](https://arxiv.org/pdf/2203.06403) has the sentence, "The spaces of (fiberwise) ω_X^N -compatible complex structures on N_X Z and of  ω_X^N -compatible Hermitian structures on (N_X Z, i_X ) are non-empty and contractible."


["empty or contractible"](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C47&q=%22empty+or+contractible%22&btnG=) (742 results)
 - I think this suggest that the author either does not accept that empty spaces are contractible, or is at least concerned that this is a point of ambiguity. Otherwise the phrase is redundant.
 - E.g., [*Ricci flow and contractibility of spaces of metrics*](https://arxiv.org/abs/1909.08710) has the sentence, "We show that the space of metrics of positive scalar curvature on any 3-manifold is either empty or contractible. "

Also,
["empty or non-contractible"](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C47&q=%22empty+or+non-contractible%22&btnG=) comes up in 10 results (mostly by the same few authors) in reference to something called the disk complex. I don't know is this is a topological space though and did not look very closely, but I thought it might be worth mentioning.

**Comment by prabau** [comment]
If you go by the definition "contractible = homotopy equivalent to a one-point space", it's obvious that the space cannot be empty. The empty set $X=\emptyset$ is not homotopy equiv. to $Y=\\{a\\}$ because by def. of homotopy equivalence, we would need maps $X\to Y$ and $Y\to X$ satisfying some conditions.  But there is no function from a non-empty set to the empty set.

**Comment by GeoffreySangston** [comment]
> If you go by the definition "contractible = homotopy equivalent to a one-point space", it's obvious that the space cannot be empty. The empty set X = ∅ is not homotopy equiv. to Y = { a } because by def. of homotopy equivalence, we would need maps X → Y and Y → X satisfying some conditions. But there is no function from a non-empty set to the empty set.
> 
> So we should go with that as the primary definition. The fact that the empty set is not contractible does not need to be part of the definition. It can be mentioned afterwards, but will follow automatically.

Do we mention that

**Comment by GeoffreySangston** [comment]
sdfd

File: theorems/T000585.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+uid: T000585
+if:
+  P000199: true
+then:
+  P000137: false
+---
+
+By definition.
```
Comment (prabau) [inline theorems/T000585.md:9]:
```suggestion
The empty space is not homotopy equivalent to a one-point space.
```

**Comment by prabau** [comment]
Looks pretty good now.  I have added one last suggestion for T585.

**Review by prabau — APPROVED** [review]

**Comment by GeoffreySangston** [comment]
> As for the git workflow, we follow a common model of working in "topic branches" off of `main`, that get reviewed and merged back into `main`. So we keep making changes on the same topic branch until everybody is satisfied. There is absolutely no need to create separate branches for each additional change while this is being worked on.

If I were to continue adding theorems related to 'contractible' (e.g., there's the unit interval, and also arbitrary products of the contractible spaces are contractible), is it better to do that in a new topic branch?

**Comment by StevenClontz** [comment]
Thanks for your contribution @GeoffreySangston! You've been invited to the repository, which means during future contributions, you can (and should) push directly to a branch of `pi-base/data` and make your PR that way, so our automated processes will make previews of your contribution available at <https://topology.pi-base.org/dev>.

**Pull request merged**